File: rust/features/src/rolling.rs
//! Rolling event-time primitives (mirrors `iap/features/rolling.py`).
//!
//! All windows are half-open intervals `(t - w, t]` in event time
//! (`exchange_ts`): a sample stamped exactly `t - w` has left the window, a
//! sample stamped `t` is inside it. Integer inputs keep integer running sums
//! (window sums are *exact*); float inputs accumulate with the same
//! add/evict ordering as the Python reference, well inside the 1e-9 golden
//! tolerance.
//!
//! All buffers are preallocated (`with_capacity`) at construction and only
//! grow if a window ever holds more samples than the initial capacity.

use std::collections::VecDeque;

/// Initial per-window sample capacity (golden vectors stay far below this).
const INIT_CAPACITY: usize = 4096;

/// Rolling sums of an `N`-tuple of `i64` values over an event-time window.
///
/// Integer sums are exact — incremental state is bit-identical to a
/// brute-force recomputation over the retained samples.
#[derive(Debug, Clone)]
pub struct RollingISum<const N: usize> {
    window: i64,
    buf: VecDeque<(i64, [i64; N])>,
    /// Running sums of each component over the live window.
    pub sums: [i64; N],
    /// Number of live samples.
    pub count: usize,
}

impl<const N: usize> RollingISum<N> {
    /// New rolling sum over `window_ns` (must be > 0).
    pub fn new(window_ns: i64) -> RollingISum<N> {
        RollingISum {
            window: window_ns,
            buf: VecDeque::with_capacity(INIT_CAPACITY),
            sums: [0; N],
            count: 0,
        }
    }

    /// Append one sample at event time `ts`, then evict expired samples.
    pub fn add(&mut self, ts: i64, vals: [i64; N]) {
        self.buf.push_back((ts, vals));
        for (sum, v) in self.sums.iter_mut().zip(vals.iter()) {
            *sum += v;
        }
        self.count += 1;
        self.trim(ts);
    }

    /// Evict samples with `ts <= now - window`.
    pub fn trim(&mut self, now: i64) {
        let cutoff = now - self.window;
        while let Some(&(ts, vals)) = self.buf.front() {
            if ts > cutoff {
                break;
            }
            for (sum, v) in self.sums.iter_mut().zip(vals.iter()) {
                *sum -= v;
            }
            self.count -= 1;
            self.buf.pop_front();
        }
    }

    /// Live samples, oldest first (brute-force verification hook).
    pub fn samples(&self) -> impl Iterator<Item = &(i64, [i64; N])> {
        self.buf.iter()
    }
}

/// Rolling sum of one `f64` value over an event-time window.
///
/// Accumulation order (add on `add`, subtract on `trim`, FIFO) matches the
/// Python reference so float drift is shared, and stays far inside the
/// golden tolerance either way.
#[derive(Debug, Clone)]
pub struct RollingFSum {
    window: i64,
    buf: VecDeque<(i64, f64)>,
    /// Running sum over the live window.
    pub sum: f64,
    /// Number of live samples.
    pub count: usize,
}

impl RollingFSum {
    /// New rolling sum over `window_ns` (must be > 0).
    pub fn new(window_ns: i64) -> RollingFSum {
        RollingFSum {
            window: window_ns,
            buf: VecDeque::with_capacity(INIT_CAPACITY),
            sum: 0.0,
            count: 0,
        }
    }

    /// Append one sample at event time `ts`, then evict expired samples.
    pub fn add(&mut self, ts: i64, val: f64) {
        self.buf.push_back((ts, val));
        self.sum += val;
        self.count += 1;
        self.trim(ts);
    }

    /// Evict samples with `ts <= now - window`.
    pub fn trim(&mut self, now: i64) {
        let cutoff = now - self.window;
        while let Some(&(ts, val)) = self.buf.front() {
            if ts > cutoff {
                break;
            }
            self.sum -= val;
            self.count -= 1;
            self.buf.pop_front();
        }
    }

    /// Live samples, oldest first (brute-force verification hook).
    pub fn samples(&self) -> impl Iterator<Item = &(i64, f64)> {
        self.buf.iter()
    }
}

/// Append-only `(ts, value)` series with at-or-before lookup and trimming
/// (mirrors the Python `TimeSeries`: compaction keeps lookups cheap while
/// retaining the newest sample at-or-before the trim horizon).
#[derive(Debug, Clone)]
pub struct TimeSeries<T: Copy> {
    ts: Vec<i64>,
    vals: Vec<T>,
    start: usize,
}

const COMPACT_AT: usize = 4096;

impl<T: Copy> Default for TimeSeries<T> {
    fn default() -> TimeSeries<T> {
        TimeSeries::new()
    }
}

impl<T: Copy> TimeSeries<T> {
    /// New empty series.
    pub fn new() -> TimeSeries<T> {
        TimeSeries {
            ts: Vec::with_capacity(INIT_CAPACITY),
            vals: Vec::with_capacity(INIT_CAPACITY),
            start: 0,
        }
    }

    /// Append a sample (timestamps must be non-decreasing).
    pub fn append(&mut self, ts: i64, val: T) {
        debug_assert!(self.ts.last().map_or(true, |&last| ts >= last));
        self.ts.push(ts);
        self.vals.push(val);
    }

    /// Latest value with sample ts <= `t`, or `None`.
    pub fn at_or_before(&self, t: i64) -> Option<T> {
        let i = self.ts[self.start..].partition_point(|&x| x <= t) + self.start;
        if i > self.start {
            Some(self.vals[i - 1])
        } else {
            None
        }
    }

    /// Most recent value, or `None` when empty.
    pub fn last(&self) -> Option<T> {
        self.vals.last().copied()
    }

    /// Number of retained samples.
    pub fn len(&self) -> usize {
        self.ts.len() - self.start
    }

    /// True when no samples are retained.
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Forget samples with ts < `min_ts`, keeping the newest at-or-before.
    pub fn trim(&mut self, min_ts: i64) {
        let i = self.ts.partition_point(|&x| x <= min_ts);
        if i > 0 {
            self.start = self.start.max(i - 1);
        }
        if self.start >= COMPACT_AT {
            self.ts.drain(..self.start);
            self.vals.drain(..self.start);
            self.start = 0;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn window_is_half_open() {
        // A sample stamped exactly t - w is OUT; one stamped t is IN.
        let mut w: RollingISum<1> = RollingISum::new(1_000);
        w.add(0, [5]);
        w.add(1_000, [7]); // trim at t=1000 evicts the t=0 sample (0 <= 0)
        assert_eq!(w.sums[0], 7);
        assert_eq!(w.count, 1);
        w.add(1_999, [1]);
        assert_eq!(w.sums[0], 8);
        w.trim(2_000); // cutoff 1000: evicts the t=1000 sample
        assert_eq!(w.sums[0], 1);
        assert_eq!(w.count, 1);
    }

    #[test]
    fn integer_sums_are_exact_vs_brute_force() {
        let mut w: RollingISum<2> = RollingISum::new(500);
        for i in 0..200i64 {
            w.add(i * 7, [i, -2 * i]);
            let brute: (i64, i64) = w
                .samples()
                .fold((0, 0), |acc, &(_, v)| (acc.0 + v[0], acc.1 + v[1]));
            assert_eq!((w.sums[0], w.sums[1]), brute);
        }
    }

    #[test]
    fn float_sum_trims_like_int_sum() {
        let mut w = RollingFSum::new(1_000);
        w.add(0, 1.5);
        w.add(500, 2.5);
        assert_eq!(w.sum, 4.0);
        w.trim(1_400); // cutoff 400: evicts the t=0 sample only
        assert_eq!(w.sum, 2.5);
        assert_eq!(w.count, 1);
    }

    #[test]
    fn timeseries_at_or_before_semantics() {
        let mut s: TimeSeries<i64> = TimeSeries::new();
        assert_eq!(s.at_or_before(100), None);
        s.append(10, 1);
        s.append(20, 2);
        s.append(20, 3); // equal timestamps allowed; latest wins
        s.append(50, 4);
        assert_eq!(s.at_or_before(9), None);
        assert_eq!(s.at_or_before(10), Some(1));
        assert_eq!(s.at_or_before(20), Some(3));
        assert_eq!(s.at_or_before(49), Some(3));
        assert_eq!(s.at_or_before(1_000), Some(4));
        assert_eq!(s.last(), Some(4));
    }

    #[test]
    fn timeseries_trim_keeps_newest_at_or_before() {
        let mut s: TimeSeries<i64> = TimeSeries::new();
        for i in 0..10 {
            s.append(i * 10, i);
        }
        s.trim(35); // keeps the ts=30 sample (newest <= 35) onward
        assert_eq!(s.at_or_before(34), Some(3));
        assert_eq!(s.at_or_before(29), None);
        assert_eq!(s.len(), 7);
    }
}

File: python/src/iap/features/rolling.py
"""Rolling event-time state primitives for the feature engine.

All windows are half-open intervals ``(t - w, t]`` in **event time**
(``exchange_ts``): a sample stamped exactly ``t - w`` has left the window,
a sample stamped ``t`` is inside it.  Every structure is O(1) amortized per
event and fully deterministic (no wall clock, no unordered iteration).

Integer inputs keep integer running sums, so incremental window sums are
*exact* (bit-identical to a brute-force recomputation); float inputs drift
only by normal IEEE accumulation, well inside the 1e-9 golden tolerance.
"""

from __future__ import annotations

from bisect import bisect_right
from collections import deque
from typing import Deque, Dict, List, Optional, Sequence, Tuple


class RollingSum:
    """Rolling sums of an n-tuple of values over an event-time window.

    ``add(ts, vals)`` appends a sample; ``trim(now)`` evicts samples with
    ``ts <= now - window``.  ``sums[i]`` is the running sum of component i,
    ``count`` the number of live samples.
    """

    __slots__ = ("window", "nvals", "buf", "sums", "count")

    def __init__(self, window_ns: int, nvals: int) -> None:
        if window_ns <= 0 or nvals <= 0:
            raise ValueError("window_ns and nvals must be positive")
        self.window = window_ns
        self.nvals = nvals
        self.buf: Deque[Tuple] = deque()
        self.sums: List = [0] * nvals
        self.count = 0

    def add(self, ts: int, vals: Sequence) -> None:
        """Append one sample at event time ``ts`` and evict expired samples."""
        self.buf.append((ts, vals))
        sums = self.sums
        for i, v in enumerate(vals):
            sums[i] += v
        self.count += 1
        self.trim(ts)

    def trim(self, now: int) -> None:
        """Evict samples with ts <= now - window."""
        cutoff = now - self.window
        buf = self.buf
        sums = self.sums
        while buf and buf[0][0] <= cutoff:
            _, vals = buf.popleft()
            for i, v in enumerate(vals):
                sums[i] -= v
            self.count -= 1


class RollingExtrema:
    """Rolling max and min of a scalar over an event-time window.

    Two monotonic deques; O(1) amortized. ``max()``/``min()`` return None
    when the window is empty.
    """

    __slots__ = ("window", "_maxq", "_minq")

    def __init__(self, window_ns: int) -> None:
        self.window = window_ns
        self._maxq: Deque[Tuple[int, float]] = deque()
        self._minq: Deque[Tuple[int, float]] = deque()

    def add(self, ts: int, val) -> None:
        maxq, minq = self._maxq, self._minq
        while maxq and maxq[-1][1] <= val:
            maxq.pop()
        maxq.append((ts, val))
        while minq and minq[-1][1] >= val:
            minq.pop()
        minq.append((ts, val))
        self.trim(ts)

    def trim(self, now: int) -> None:
        cutoff = now - self.window
        while self._maxq and self._maxq[0][0] <= cutoff:
            self._maxq.popleft()
        while self._minq and self._minq[0][0] <= cutoff:
            self._minq.popleft()

    def max(self):
        return self._maxq[0][1] if self._maxq else None

    def min(self):
        return self._minq[0][1] if self._minq else None


class TimeSeries:
    """Append-only (ts, value) series with at-or-before lookup and trimming.

    ``at_or_before(t)`` returns the latest value with sample ts <= t (None if
    none).  ``trim(min_ts)`` forgets samples older than min_ts (amortized via
    periodic compaction so lookups stay O(log n) on the retained span).
    """

    __slots__ = ("_ts", "_vals", "_start")

    _COMPACT_AT = 4096

    def __init__(self) -> None:
        self._ts: List[int] = []
        self._vals: List = []
        self._start = 0

    def append(self, ts: int, val) -> None:
        if self._ts and ts < self._ts[-1]:
            raise ValueError("TimeSeries requires non-decreasing timestamps")
        self._ts.append(ts)
        self._vals.append(val)

    def at_or_before(self, t: int):
        """Latest value with ts <= t, or None."""
        i = bisect_right(self._ts, t, self._start)
        return self._vals[i - 1] if i > self._start else None

    def first_ts(self) -> Optional[int]:
        return self._ts[self._start] if self._start < len(self._ts) else None

    def last(self):
        return self._vals[-1] if self._vals else None

    def __len__(self) -> int:
        return len(self._ts) - self._start

    def trim(self, min_ts: int) -> None:
        """Forget samples with ts < min_ts (keeps the newest at-or-before)."""
        i = bisect_right(self._ts, min_ts)
        if i > 0:
            self._start = max(self._start, i - 1)
        if self._start >= self._COMPACT_AT:
            del self._ts[: self._start]
            del self._vals[: self._start]
            self._start = 0


class RollingKeyCount:
    """Rolling per-key event counts over an event-time window (venue shares)."""

    __slots__ = ("window", "buf", "counts", "total")

    def __init__(self, window_ns: int) -> None:
        self.window = window_ns
        self.buf: Deque[Tuple[int, int]] = deque()
        self.counts: Dict[int, int] = {}
        self.total = 0

    def add(self, ts: int, key: int) -> None:
        self.buf.append((ts, key))
        self.counts[key] = self.counts.get(key, 0) + 1
        self.total += 1
        self.trim(ts)

    def trim(self, now: int) -> None:
        cutoff = now - self.window
        while self.buf and self.buf[0][0] <= cutoff:
            _, key = self.buf.popleft()
            c = self.counts[key] - 1
            if c:
                self.counts[key] = c
            else:
                del self.counts[key]
            self.total -= 1

    def shares(self) -> List[float]:
        """Per-key count shares, iterated in sorted key order."""
        if not self.total:
            return []
        tot = float(self.total)
        return [self.counts[k] / tot for k in sorted(self.counts)]


class SessionProfile:
    """Expanding minute-of-day session profile built from the data itself.

    **Bound / regime caveat (documented, round-3)**: the profile is
    EXPANDING and never forgets.  Over weeks the normalizer converges to a
    long-run mean and stops tracking the current regime, so
    ``norm_*_m5_v1`` measures "vs the whole history", not "vs recent
    comparable sessions".  Memory is bounded (288 buckets x metrics), so
    this is a modelling choice, not a leak: a decayed variant is available
    via ``decay`` (an exponential-forgetting factor applied to the running
    count and sum at each update; ``decay = 1.0``, the default, is the
    pinned expanding behaviour and the only one the goldens cover).

    Buckets are 5 minutes of the SESSION-LOCAL day (0..287).  For each
    (metric, bucket) the
    profile keeps an expanding count and sum.  ``prior(metric, bucket)``
    returns the (count, mean) accumulated *before* the current observation —
    readers must call ``prior`` before ``update`` so normalization never uses
    the value being normalized (no lookahead).
    """

    BUCKETS = 288  # 5-minute buckets per session-local day
    MIN_OBS = 10  # observations required before a normalization is valid

    __slots__ = ("_count", "_sum", "decay")

    def __init__(self, metrics: Sequence[str], decay: float = 1.0) -> None:
        if not 0.0 < decay <= 1.0:
            raise ValueError("decay must be in (0, 1]")
        self.decay = float(decay)
        self._count: Dict[str, List[float]] = {
            m: [0.0] * self.BUCKETS for m in metrics
        }
        self._sum: Dict[str, List[float]] = {
            m: [0.0] * self.BUCKETS for m in metrics
        }

    @staticmethod
    def bucket_of(ts_ns: int, utc_offset_s: int = 0) -> int:
        """5-minute-of-day bucket for an event timestamp.

        ``utc_offset_s`` is the venue's UTC offset at that instant
        (``InstrumentContext.clock``): buckets are keyed in SESSION-LOCAL
        time so a DST shift moves the whole profile with the venue
        (API_FEATURES §3).  The default 0 keeps pure-UTC callers exact.
        """
        sec_of_day = (ts_ns // 1_000_000_000 + utc_offset_s) % 86_400
        return int(sec_of_day // 300)

    def prior(self, metric: str, bucket: int) -> Tuple[float, float]:
        """(count, mean) accumulated so far for (metric, bucket); mean=0 if empty."""
        c = self._count[metric][bucket]
        s = self._sum[metric][bucket]
        return c, (s / c if c else 0.0)

    def update(self, metric: str, bucket: int, value: float) -> None:
        """Fold one observation into the profile (call after ``prior``).

        With ``decay < 1`` the running count and sum are scaled first, so
        the effective memory is ``1 / (1 - decay)`` observations per bucket.
        """
        d = self.decay
        if d != 1.0:
            self._count[metric][bucket] *= d
            self._sum[metric][bucket] *= d
        self._count[metric][bucket] += 1.0
        self._sum[metric][bucket] += value
